packages/subgraph/src/util.ts

```diff
@@ -29,3 +115,374 @@ export function loadOrCreateAccount(accountId: string): Account {
+
+  // Unrealized PnL calculation
+  let oldUnrealizedETHCost = position.unrealizedETHUnitCost.times(
+    position.currentETHAmount
+  );
+  position.currentETHAmount = position.currentETHAmount.plus(amount);
+  // = 0 none
+  if (
+    position.currentOSQTHAmount.equals(ZERO_BD) &&
+    position.currentETHAmount.equals(ZERO_BD)
+  ) {
+    position = initPosition(userAddr, position);
+  } else if (!position.currentETHAmount.equals(ZERO_BD)) {
+    position.unrealizedETHUnitCost = oldUnrealizedETHCost
+      .plus(amount.times(usdcPrices[1]))
+      .div(position.currentETHAmount);
+  }
+
+  position.save();
+}
+
+// buy: amount > 0
+// sell amount < 0
+export function buyOrSellSQTH(userAddr: string, amount: BigDecimal): void {
+  if (amount.equals(ZERO_BD)) return;
+  let osqthPrices = getoSQTHETHPrice();
+
+  let position = loadOrCreatePosition(userAddr);
+
+  // When position side chages, reset PnLs and calculate with remaining amount
+  let newAmount = position.currentOSQTHAmount.plus(amount);
+  if (position.currentOSQTHAmount.times(newAmount).lt(ZERO_BD)) {
+    position.currentOSQTHAmount = ZERO_BD;
+    position.realizedOSQTHAmount = ZERO_BD;
+    amount = newAmount;
+  }
+
+  // Buy
+  if (amount.gt(ZERO_BD)) {
+    let oldBoughtAmount = position.currentOSQTHAmount.plus(
+      position.realizedOSQTHAmount
+    );
+    let oldRealizedOSQTHCost =
+      position.realizedOSQTHUnitCost.times(oldBoughtAmount);
+
+    position.realizedOSQTHUnitCost = oldRealizedOSQTHCost
+      .plus(amount.times(osqthPrices[3]))
+      .div(oldBoughtAmount.plus(amount));
+  }
+
+  // Sell
+  if (amount.lt(ZERO_BD)) {
+    let absAmount = amount.neg();
+    let oldRealizedOSQTHGain = position.realizedOSQTHUnitCost.times(
+      position.realizedOSQTHUnitGain
+    );
+
+    position.realizedOSQTHAmount = position.realizedOSQTHAmount.plus(absAmount);
+    position.realizedOSQTHUnitGain = oldRealizedOSQTHGain
+      .plus(absAmount.times(osqthPrices[3]))
+      .div(position.realizedOSQTHAmount);
+  }
+
+  // Unrealized PnL calculation
+  let oldUnrealizedOSQTHCost = position.unrealizedOSQTHUnitCost.times(
+    position.currentOSQTHAmount
+  );
+
+  position.currentOSQTHAmount = position.currentOSQTHAmount.plus(amount);
+  // = 0 none
+  if (
+    position.currentOSQTHAmount.equals(ZERO_BD) &&
+    position.currentETHAmount.equals(ZERO_BD)
+  ) {
+    position = initPosition(userAddr, position);
+  } else if (!position.currentOSQTHAmount.equals(ZERO_BD)) {
+    position.unrealizedOSQTHUnitCost = oldUnrealizedOSQTHCost
+      .plus(amount.times(osqthPrices[3]))
+      .div(position.currentOSQTHAmount);
+  }
+
+  position.save();
+}
+
+// buy: amount > 0
+// sell amount < 0
+export function buyOrSellLPETH(userAddr: string, amount: BigDecimal): void {
+  if (amount.equals(ZERO_BD)) return;
+  let usdcPrices = getETHUSDCPrice();
+  let position = loadOrCreateLPPosition(userAddr);
+
+  // When position side chages, reset PnLs and calculate with remaining amount
```

Comment:
We wont have negative amount in lp. Plus, we dont need the position type check for lp, we can just clear the whole position when they both equals zero

Reply:
I got it.